lib/block_view_message.go

```diff
@@ -825,95 +949,189 @@ func (bav *UtxoView) _connectMessagingGroup(
+					"_connectMessagingGroup: GroupOwner cannot mute herself (%v).", existingEntry.GroupOwnerPublicKey[:])
+			}
+			// Add member to newMuteList. We iterate over newMuteList to prevent mute list entries with duplicate public keys.
+			for _, addedMutedMember := range newMuteList {
+				// Check if member is already muted, in such case we error.
+				if reflect.DeepEqual(addedMutedMember.GroupMemberPublicKey[:], newlyMutedMember.GroupMemberPublicKey[:]) {
+					return 0, 0, nil, errors.Wrapf(RuleErrorMessagingMemberAlreadyMuted,
+						"_connectMessagingGroup: Cannot mute member that is already muted (%v).", newlyMutedMember.GroupMemberPublicKey[:])
+				}
+			}
+			newMuteList = append(newMuteList, newlyMutedMember)
 		}
-		// The messagingMember can't be already added to the list of existing members.
-		if _, exists := existingMembers[*messagingMember.GroupMemberPublicKey]; exists {
-			return 0, 0, nil, errors.Wrapf(
-				RuleErrorMessagingMemberAlreadyExists, "_connectMessagingGroup: "+
-					"Error, messagingMember already exists (%v)", messagingMember.GroupMemberPublicKey[:])
+		// Set the messaging members to the existing members since we won't be changing them.
+		newMessagingMembers = existingEntry.MessagingGroupMembers
+
+	case MessagingGroupOperationUnmuteMembers:
+		// Unmuting members assumes the group was already created.
+		if existingEntry == nil || existingEntry.isDeleted {
+			return 0, 0, nil, errors.Wrapf(RuleErrorMessagingGroupDoesntExist,
+				"_connectMessagingGroup: Can't mute members for a non-existing group")
+		}
+		for _, member := range txMeta.MessagingGroupMembers {
+			isUnmuteValid := false
+			// re-add all muted members except for the member that we're trying to unmute.
+			newMuteList = []*MessagingGroupMember{}
+			for _, toUnmute := range existingEntry.MuteList {
+				// if member not in MuteList, add member to the tempMuteList
+				if !reflect.DeepEqual(toUnmute.GroupMemberPublicKey[:], member.GroupMemberPublicKey[:]) {
+					newMuteList = append(newMuteList, toUnmute)
+				} else {
+					// if member IS in MuteList, then unmute txn is valid, and we continue looping to add remaining members to newMuteList.
+					isUnmuteValid = true
+				}
+			}
+			// If invalid unmuting, then check why:
+			// 1. GroupOwner unmuting herself
+			// 2. Member already unmuted
+			// 3. Member does not exist in group hence cannot unmute nonexistent member
+			if !isUnmuteValid {
+				// 1. GroupOwner unmuting herself is invalid because GroupOwner can never be muted in the first place
+				if reflect.DeepEqual(member.GroupMemberPublicKey[:], existingEntry.GroupOwnerPublicKey[:]) {
+					return 0, 0, nil, errors.Wrapf(RuleErrorMessagingGroupOwnerUnmutingSelf,
+						"_connectMessagingGroup: GroupOwner cannot mute herself (%v).", existingEntry.GroupOwnerPublicKey[:])
+				}
+				// 2. Member already unmuted
+				for _, entryMembers := range existingEntry.MessagingGroupMembers {
```

Comment:
we should try to remove these nested loops as much as possible

Reply:
the O(nlogn) thingy gets rid of them. This one in particular is just to come up with a better validation error, and since we will error regardless if we get into this the `if !isUnmuteValid` it makes the whole thing O(nlogn) still